Add Global React context

mattchx opened the pull request:
**Completed**
- applied React context to pull user data from BE
- currently displaying user name and profile photo using context

**To Do**
- integrate Auth context with User context
- apply context to different parts of the app as required

**EDIT:** Global context for user data and auth is fully functional

mattchx committed:
set up global context and added new route to get user data

mattchx committed:
context is working for userData

- user data is retrieved from DB and stored in global context
- custom useContext hooks allow data to be grabbed by components
- currently working for profile photos in NavBar and EventTypesTab

mattchx committed:
displayed user name using context

alvyjudy commented:
I think we should place the provider in `App.js` instead because there are other components also needing access to the context data. That's also where other contexts are placed as well (MuiThemeProvider and BrowserRouter).

kozaktar commented on client/src/providers/Context.js:
```diff
@@ -0,0 +1,52 @@
+    axios
+      .get('/api/user/data')
+      .then((res) => {
+        console.log(res.data);
```
remove console.log after testing is done

mattchx replied:
sure!

kozaktar commented on client/src/components/Header/NavBar.js:
```diff
@@ -39,6 +41,9 @@ export default function NavBar() {
 
+  const user = useUserContext()
+  const { name, photoUrl } = user
+
```
can be simplified to:
 const { name, photoUrl } = useUserContext();

mattchx replied:
Good point

kozaktar reviewed (commented):
Looks really good Matt!

mattchx committed:
Context now includes User Auth

- Waiting for Dev changes to be approved then will merge dev in this branch

mattchx commented:
> I think we should place the provider in `App.js` instead because there are other components also needing access to the context data. That's also where other contexts are placed as well (MuiThemeProvider and BrowserRouter).

Absolutely, I can update the draft.

mattchx committed:
Merge branch 'dev'

Conflicts:
	client/src/components/Header/NavBar.js
	client/src/components/SelectTabs/EventTypesTab.js
	client/src/pages/Home.js

mattchx committed:
code clean up

mattchx committed:
transitioned AuthProvider into UserContext

- Wrapped UserContextProvider around App.js
- cleaned up code
- replaced AuthProvider.js with Context.js

mattchx committed:
misc clean up

mattchx committed:
added setUserUpdated to Context

mattchx committed:
Update ImageUploader.js

mattchx committed:
Update imageUploaderAPI.js

mattchx committed:
small correction

mattchx committed:
Added photoUrl to user schema

mattchx committed:
testing setUserHasUpdated

mattchx commented:
Context is working. 
Currently issue where I cannot get context to update and rerender as user information is updated. 
I believe there are two options. 
1) To update the user state object directly by passing the updated variable (eg photoUrl) back into context. 
2) To trigger a new get request in context causing the user state object to be updated. (I plan to implement this)
- Bascially I have passed custom hooks that pass true back into context which then tiggers a useEffect to start a get request

Please let me know if you have ideas how to correct this or a better method. 
Thanks!

alvyjudy commented on client/src/providers/Context.js:
```diff
@@ -0,0 +1,71 @@
+  return setAuthenticated;
+};
+// Get User Data
+export const useUserContext = () => {
```
I think we should name this `useUserData` instead. 
In the comment let's also briefly outline what the data would include. Something like this:
```
// Usage: 
// const userData = useUserData(); 
// {_id, email, name, googleCredentials, photoUrl, URL, stripeId, subscribed, subscriptionId, ...}
```

which brings me to the question regarding how much of this data we want to expose to the frontend 🧐

mattchx replied:
That makes sense. I can make the name change.

alvyjudy commented on client/src/providers/Context.js:
```diff
@@ -0,0 +1,71 @@
+  return useContext(UserContext).userData;
+};
+// Set User Data has Updated
+export const useSetUserHasUpdated = () => {
```
If I understand correctly with the set up, the child component calls a function which triggers the context to update on its own and renders the child with updated props. In that case, we should name this `useUpdateData` and have it return a function, which can be invoked without any argument.
```
const useUpdateData= ()=>{
    // Usage: 
    // const updateData = useUpdateData()
    // updateData()
    const setUserHasUpdated = useContext(UserContext).setUserHasUpdated;
    const updateData = () => {setUserHasUpdated(true)}
    return updateData
}
```

mattchx replied:
It's must cleaner without needing to be passed a boolean argument in the child component. Thanks!

alvyjudy commented on client/src/providers/Context.js:
```diff
@@ -0,0 +1,71 @@
+  const [userHasUpdated, setUserHasUpdated] = useState(false);
+
+  // Gets User Data from DB
+  useEffect(() => {
+    axios
+      .get("/api/user/data")
+      .then((res) => {
+        setUserData(res.data);
+      })
+      .catch((err) => console.log(err));
+  }, []);
```
I'm not sure why this block is necessary. Wouldn't the second useEffect fire on page load anyways?

mattchx replied:
Yes, it does. I have removed the first useEffect.

alvyjudy reviewed (commented):
Your code does seem to work for me actually. Could you elaborate on the issue you are experiencing? I left a few comments and suggestions based on my understanding of your set up. Let me know if they don't make sense because I might have misinterpreted your code. We can discuss them on a call tomorrow!

A big issue I noticed is too many re-rendering. I haven't come up with an idea where things could be optimized but worth looking into.

dproc96 commented on client/src/providers/Context.js:
```diff
@@ -0,0 +1,71 @@
+        setUserData(res.data);
+      })
+      .catch((err) => console.log(err));
+    setUserHasUpdated(false);
```
The way you've written this, it will make the API request fire twice. The reason is because useEffect runs every time the value of its dependencies changes so it goes like this:

1. userUpdated is set to true
2. useEffect detects a change
3. The API request is fired
4. userUpdated is changed to false
5. useEffect detects a change
6. The API request is fired
7. userUpdated is set to false, but because it was already false the useEffect doesn't detect a change so it doesn't run a third time

dproc96 commented on server/routes/imageUploaderAPI.js:
```diff
@@ -1,37 +1,54 @@
+        const AwsUrl = `https://${BUCKET}.s3.amazonaws.com/${req.file.key}`;
+        db.User.updateOne({ _id: req.userId }, { $set: { photoUrl: AwsUrl } })
+          .then((response) => {
+            res.send(response);
```
You're sending a response twice in this controller function which causes the app to crash

mattchx replied:
For some reason I thought I must send a response when the DB was updated. But then I realized that only one response is needed.

mattchx committed:
changed name from useUserContext to useUserData

mattchx committed:
removed extra useEffect (for data fetching)

mattchx committed:
useSetUserHasUpdated name is changed to useUpdateData

alvyjudy commented on server/routes/imageUploaderAPI.js:
```diff
@@ -1,37 +1,54 @@
 // Create S3 service object
 const s3 = new AWS.S3();
-const FN = Date.now().toString() + '.png'
+const FN = Date.now().toString() + ".png";
```
This line is why image does not update!😜🧐

mattchx replied:
Thanks for your help!

mattchx committed:
fixed re-render of photo on upload

- put filename(FN) inside the upload function in imageUploaderAPI

sina-jamshidi commented on client/src/components/UploadModal/ImageUploader.js:
```diff
@@ -88,22 +86,28 @@ export default function UploadDialog(props) {
       setMessage(true);
-      await setTimeout(() => {
+      
+      // const {fileName} = res.data
+      // const newPhotoUrl = `https://${BUCKET}.s3.amazonaws.com/${fileName}`
```
let's remove commented code and console.logs

sina-jamshidi commented on server/routes/imageUploaderAPI.js:
```diff
@@ -1,37 +1,55 @@
+          msg: "Your image was uploaded sucessfully",
+          fileName: req.file.key,
+        });
+        const AwsUrl = `https://${BUCKET}.s3.amazonaws.com/${req.file.key}`;
+        db.User.updateOne({ _id: req.userId }, { $set: { photoUrl: AwsUrl } })
+          .then((response) => {
+            res.send(response);
+          })
```
👍

sina-jamshidi commented:
> Context is working.
> Currently issue where I cannot get context to update and rerender as user information is updated.
> I believe there are two options.
> 
> 1. To update the user state object directly by passing the updated variable (eg photoUrl) back into context.
> 2. To trigger a new get request in context causing the user state object to be updated. (I plan to implement this)
> 
> * Bascially I have passed custom hooks that pass true back into context which then tiggers a useEffect to start a get request
> 
> Please let me know if you have ideas how to correct this or a better method.
> Thanks!

I'm a little worried that your way of fixing this is too hacky and technically a forced update, which is a React antipattern. Can you elaborate on the re-render issue? What is a specific case where it doesn't work?

mattchx commented:
> > Context is working.
> > Currently issue where I cannot get context to update and rerender as user information is updated.
> > I believe there are two options.
> > 
> > 1. To update the user state object directly by passing the updated variable (eg photoUrl) back into context.
> > 2. To trigger a new get request in context causing the user state object to be updated. (I plan to implement this)
> > 
> > 
> > * Bascially I have passed custom hooks that pass true back into context which then tiggers a useEffect to start a get request
> > 
> > Please let me know if you have ideas how to correct this or a better method.
> > Thanks!
> 
> I'm a little worried that your way of fixing this is too hacky and technically a forced update, which is a React antipattern. Can you elaborate on the re-render issue? What is a specific case where it doesn't work?

The render was fixed by moving the profile image filename creation inside the upload function in the BE. 
You mean option 1 (see above) would be less hacky?

mattchx committed:
code clean up

mattchx committed:
corrected formatting and one warning

mattchx committed:
Merge branch 'dev' into add-global-context

mattchx committed:
code clean up

removed second res.send(response)

mattchx committed:
Changed context to update by returning user object rather than triggering a new request

mattchx committed:
Merge branch 'add-global-context' of https://github.com/hatchways/team-yugioh into add-global-context

mattchx committed:
Added disabled button feature to photo uploader

dproc96 commented on server/routes/imageUploaderAPI.js:
```diff
@@ -42,13 +43,12 @@ router.post("/api/image-upload", auth, (req, res) => {
+          { _id: req.userId },
+          { $set: { photoUrl: AwsUrl } }
+        ).catch((error) => {
+          res.send(error);
```
I think this error handling would still crash the server if it came up, maybe waiting to send the response until after updating the DB instead of before would be a better way to handle it

mattchx replied:
Thanks for comments Dan! I did correct the error handling and also simplifed the logic in the request.

dproc96 reviewed (commented):
One small thing I noticed in the API

mattchx committed:
corrected BE error handling and simplified logic on BE/FE

dproc96 commented on server/routes/imageUploaderAPI.js:
```diff
@@ -1,37 +1,51 @@
+        ).catch((error) => {
+          res.send(error);
+        });
+        res.status(200).send({ awsUrl });
```
You'll want to move this to a .then() on Line 45 to solve potential async issuea

mattchx replied:
I did ` .then((res) =>  res.status(200).send({ awsUrl })` originally but caused the app to crash.
So now I've changed it to this: `.then(() => uploadRes.status(200).send({ awsUrl }))` which is working.

mattchx committed:
fixed error async issue

dproc96 commented on server/routes/imageUploaderAPI.js:
```diff
@@ -1,37 +1,50 @@
+        res.status(400).send("There was a error with your upload paramaters");
       } else {
-        res.status(200).send({msg:'Your image was uploaded sucessfully', fileName : req.file.key});
+        const uploadRes = res;
```
I don't think it's necessary to rename the response here, the reason your original solution wasn't working was because adding `res` as an argument in the .then() reassigned the value

mattchx replied:
That makes sense!

mattchx committed:
response handling simplification

mattchx committed:
misc change

mattchx committed:
current profile photo is previewed in modal by default

mattchx committed:
updated upload button styling

mattchx committed:
Merge branch 'dev' into add-global-context

mattchx committed:
Merge branch 'dev' into add-global-context
